File: testing/MLDB-283-embedding-nearest-neighbours.js
// This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.

/* MLDB-283 Test of nearest neighbours. */

function assertEqual(expr, val, msg)
{
    if (expr == val)
        return;
    if (JSON.stringify(expr) == JSON.stringify(val))
        return;

    plugin.log("expected", val);
    plugin.log("received", expr);

    throw "Assertion failure: " + msg + ": " + JSON.stringify(expr)
        + " not equal to " + JSON.stringify(val);
}

var dataset_config = {
    'type'    : 'embedding',
    'id'      : 'test',
};

var dataset = mldb.createDataset(dataset_config)

var ts = new Date();

function recordExample(row, x, y)
{
    dataset.recordRow(row, [ [ "x", x, ts ], ["y", y, ts] ]);
}

// Very simple linear regression, with x = y
recordExample("ex1", 0, 0);
recordExample("ex2", 0, 1);
recordExample("ex3", 1, 0);
recordExample("ex4", 1, 1);

dataset.commit()

var sql_func_res = mldb.put("/v1/functions/nn", {
    type: 'nearest.neighbors',
    params: {
        'dataset': {id: 'test', type: "embedding"}
    }
});

mldb.log(sql_func_res);


var res1 = mldb.get("/v1/datasets/test/routes/neighbours", {x:0.5,y:0.5}).json;

var expected1 = [
   [ "ex1", "397de880d5f0376e", 0.7071067690849304 ],
   [ "ex2", "ed64a202cef7ccf1", 0.7071067690849304 ],
   [ "ex3", "418b8ce19e0de7a3", 0.7071067690849304 ],
   [ "ex4", "213ca5902e95224e", 0.7071067690849304 ]
];

assertEqual(mldb.diff(expected1, res1, false /* strict */), {},
            "Output was not the same as expected output");


var res1_sql = mldb.query("select nn({coords: [0.5, 0.5]})[neighbors] as *");
mldb.log(res1_sql);
for(var i=0; i<res1_sql[0]["columns"].length; i++) {
    var elem = res1_sql[0]["columns"][i];
    assertEqual(elem[0], expected1[i][0]);
    assertEqual(elem[1], expected1[i][2]);
}


var res2 = mldb.get("/v1/datasets/test/routes/neighbours", {x:0.1,y:0.2}).json;

if (res2[0][0] != "ex1"
    || res2[1][0] != "ex2"
    || res2[2][0] != "ex3"
    || res2[3][0] != "ex4")
    throw "Row names are wrong";

if (res2[0][2] >= res2[1][2]
    || res2[1][2] >= res2[2][2]
    || res2[2][2] >= res2[3][2])
    throw "Row order is wrong";




var expected2 = [
   [ "ex1", "397de880d5f0376e", 0.22360680997371674 ],
   [ "ex2", "ed64a202cef7ccf1", 0.8062257766723633 ],
   [ "ex3", "418b8ce19e0de7a3", 0.9219543933868408 ],
   [ "ex4", "213ca5902e95224e", 1.2041594982147217 ]
];

if (JSON.stringify(expected2) != JSON.stringify(res2)) {
    plugin.log(JSON.stringify(expected2));
    plugin.log(JSON.stringify(res2));
    assertEqual(mldb.diff(expected2, res2, false /* strict */), {},
                "Output was not the same as expected output");
}

plugin.log("result of query 2", res2);

// MLDB-509
var res3 = mldb.get("/v1/datasets/test/routes/rowNeighbours", {row: "ex1"}).json;

var expected3 = [
   [ "ex1", "397de880d5f0376e", 0 ],
   [ "ex2", "ed64a202cef7ccf1", 1 ],
   [ "ex3", "418b8ce19e0de7a3", 1 ],
   [ "ex4", "213ca5902e95224e", 1.4142135381698608 ]
]

plugin.log("result of query 3", res3);

if (JSON.stringify(expected3) != JSON.stringify(res3)) {
    plugin.log(JSON.stringify(expected3));
    plugin.log(JSON.stringify(res3));
    assertEqual(mldb.diff(expected3, res3, false /* strict */), {},
                "Output was not the same as expected output");
}


var res3_sql = mldb.query("select nn({coords: 'ex1'})[neighbors] as *");
mldb.log(res3_sql);
for(var i=0; i<res3_sql[0]["columns"].length; i++) {
    var elem = res3_sql[0]["columns"][i];
    assertEqual(elem[0], expected3[i][0]);
    assertEqual(elem[1], expected3[i][2]);
}

var res4_sql = mldb.query("select nn({coords: 'ex1', num_neighbours:2})[neighbors] as *");
mldb.log(res4_sql);
assertEqual(res4_sql[0]["columns"].length, 2);

var res5_sql = mldb.query("select nn({coords: 'ex1', num_neighbours:2, max_distance:0.5})[neighbors] as *");
mldb.log(res5_sql);
assertEqual(res5_sql[0]["columns"].length, 1);
assertEqual(res5_sql[0]["columns"][0][0], 'ex1');

"success"

File: testing/MLDB-283-embedding-nearest-neighbours.py
#
# MLDB-283-embedding-nearest-neighbours.py
# 2016-03-14
# This file is part of MLDB. Copyright 2016 mldb.ai inc. All rights reserved.
#


import unittest

mldb = mldb_wrapper.wrap(mldb) # noqa

class Mldb283Test(MldbUnitTest):

    @classmethod
    def setUpClass(self):
        # create a dummy dataset
        ds = mldb.create_dataset({ "id": "test", "type": "embedding" })
        
        ts = 0
        ds.record_row("ex1", [ [ "x", 0, ts ], ["y", 0, ts]])
        ds.record_row("ex2", [ [ "x", 0, ts ], ["y", 1, ts]])
        ds.record_row("ex3", [ [ "x", 1, ts ], ["y", 0, ts]])
        ds.record_row("ex4", [ [ "x", 1, ts ], ["y", 1, ts]])

        ds.commit()

        # create nn function
        mldb.put("/v1/functions/nn", {
            "type": 'embedding.neighbors',
            "params": {
                'dataset': 'test'
            }
        })

    def test_select(self):
        result = mldb.query("select nn({coords: {x:0.5, y:0.5}})[distances] as *")
        expected = [0.7071067690849304,
                    0.7071067690849304,
                    0.7071067690849304,
                    0.7071067690849304 ]

        for i, dist in enumerate(result[1][1:5]):
            # testing up to a given precision since the value varies depending 
            # on the CPU capabilities
            self.assertAlmostEqual(dist, expected[i], 5)
        
        result = mldb.query("select nn({coords: {x:0.1, y:0.2}})[distances] as *")
        expected =  [0.22360680997371674,
                     0.8062257766723633,
                     0.9219543933868408,
                     1.2041594982147217]
        
        for i, dist in enumerate(result[1][1:5]):
            # testing up to a given precision since the value varies depending 
            # on the CPU capabilities
            self.assertAlmostEqual(dist, expected[i], 5)
    
    def test_select_row(self):
        # MLDB-509
        self.assertTableResultEquals(
            mldb.query("select nn({coords: 'ex1'})[distances] as *"),
            [
                ["_rowName", "ex1", "ex2", "ex3", "ex4"],
                ["result",  0, 1, 1, 1.4142135381698608]
            ]
        )

    def test_select_params(self):
        self.assertTableResultEquals(
            mldb.query("select nn({coords: 'ex1', numNeighbors:2})[distances] as *"),
            [
                ["_rowName", "ex1", "ex2"],
                ["result",  0, 1]
            ]
        )
        
        self.assertTableResultEquals(
            mldb.query("select nn({coords: 'ex1', numNeighbors:2, maxDistance:0.5})[distances] as *"),
            [
                ["_rowName", "ex1"],
                ["result",  0]
            ]
        )

mldb.run_tests()

